### DISCUSSION TITLE
Need advice on tensorflow embedding layers.

### DISCUSSION THREAD

**Discussion by 0as1s** [horovod/horovod#2802] [Q&A]
**Is your feature request related to a problem? Please describe.**

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

HI! May I ask if Horovod provides any optimization on tensorflow's embdding layers. I have a very large embedding layer whose size is 30M * 8. It seems on every step all parameters of the embedding layer are synced among ranks, causing a very high communication cost. Does Horovod have any optimization for this scenario? Or can I tell Horovod to only sync changed gradients? Thanks a lot. 
![Image Pasted at 2021-3-29 15-35](https://user-images.githubusercontent.com/8215822/112806642-98102600-90a9-11eb-9214-30dde59c5bb9.png)

**Comment by tgaddair** [comment]
Hey @0as1s, Horovod has a couple settings you can play around with for embeddings. One of them is the param `sparse_as_dense` which you can provide to the `DistributedOptimizer`. This should improve performance in cases where the number of affected params in the embedding is very high. By default, Horovod will do an allgather on the indices and values, which is appropriate if your gradients are sparse. But because the HorovodAllgather cost here is very high, I would suggest trying `sparse_as_dense` and see how that does in comparison.

**Reply by 0as1s** [reply]
Hi @tgaddair, thanks a lot for your reply! Let me try sparse_as_dense tomorrow. You mean by default Horovod only synchronizes the updated parameters of the embedding layer? And according to the figure, you guessed that a lot of parameters of the embedding layer were updated so sparse_as_dense may reduce the time for allreduce. Please correct me if I'm wrong. Thanks again!

**Reply by tgaddair** [reply]
Yes, exactly. If the Allgather is taking up a lot of time, it's usually because there are so many params that were updated that it would be more efficient to just do an Allreduce.

**Reply by 0as1s** [reply]
<img width="1771" alt="original" src="https://user-images.githubusercontent.com/8215822/112933026-9b5eec80-9151-11eb-968c-c1f329a948ae.png">
<img width="1771" alt="original" src="https://user-images.githubusercontent.com/8215822/112933682-e7f6f780-9152-11eb-8db3-fd408a8fa11f.png">


Hi @tgaddair, thank you for your answer! I'm sorry previously  I did not notice that CUPTI was missing and in the previous figure, some ops were omitted. I reran the TensorFlow timeline for using sparse_as_dense or not after fixing this issue.  May I ask, do the two figures suggest that computing and communication are running parallelly? If so, communication may not be the bottleneck for my training. The setting for them is one machine with 3 GPUs.

Thanks a lot!

**Comment by 0as1s** [comment]
Hi @tgaddair, thanks again for your help! I finally found the bottleneck. We were using Adam optimizer before and it has a variable for each parameter in the TensorFlow graph. Therefore, on each batch, a lot of updates were made to synchronize the Adam optimizers on different ranks. Luckily, TensorFlow provides lazy Adam optimizer in tf.contrib.opt (for tf 1.x) and tf addons (for tf 2.x). It brings 3x to 4x speedup.
![lazy_adam](https://user-images.githubusercontent.com/8215822/112962623-27d0d580-9179-11eb-9d7e-50673d0f576a.png)
